File: backend/src/main/java/com/project/backend/features/customer/mapper/CustomerTransactionMapper.java
package com.project.backend.features.customer.mapper;

import org.springframework.stereotype.Component;

import com.project.backend.common.dayrule.dto.DayRule;
import com.project.backend.common.dayrule.utils.DayRuleUtils;
import com.project.backend.features.customer.dto.CustomerTransactionClosingRequest;
import com.project.backend.features.customer.dto.CustomerTransactionRequest;
import com.project.backend.features.customer.dto.CustomerTransactionResponse;
import com.project.backend.features.customer.entity.CustomerTransaction;
import com.project.backend.features.customer.enums.CustomerPaymentStatus;

@Component
public class CustomerTransactionMapper {

    public CustomerTransaction toEntity(
            Long customerId,
            CustomerTransactionRequest request
    ) {
        CustomerTransaction entity = new CustomerTransaction();
        entity.setCustomerId(customerId);
        apply(entity, request);
        return entity;
    }

    public void apply(
            CustomerTransaction entity,
            CustomerTransactionRequest request
    ) {
        entity.setTargetMonth(request.targetMonth());

        applyClosingDayRule(entity, request.closingDayRule());
        applyPaymentDayRule(entity, request.paymentDayRule());

        entity.setBillingAmount(request.billingAmount());
        entity.setExpectedPaymentDate(request.expectedPaymentDate());
        entity.setConfirmedPaymentDate(request.confirmedPaymentDate());

        entity.setPaidAmount(request.paidAmount());
        entity.setFee(request.fee());
        entity.setOffsetAmount(request.offsetAmount());
        entity.setTotalAmount(request.totalAmount());

        entity.setPaymentStatus(
                request.paymentStatus() == null
                        ? CustomerPaymentStatus.UNPAID
                        : request.paymentStatus()
        );

        entity.setNote(request.note());
    }

    public void applyFromClosing(
            CustomerTransaction entity,
            CustomerTransactionClosingRequest request
    ) {
        entity.setCustomerId(request.customerId());
        entity.setTargetMonth(request.targetMonth());

        applyClosingDayRule(entity, request.closingDayRule());
        applyPaymentDayRule(entity, request.paymentDayRule());

        entity.setBillingAmount(request.billingAmount());
        entity.setExpectedPaymentDate(request.expectedPaymentDate());

        if (entity.getPaidAmount() == null) {
            entity.setPaidAmount(0);
        }

        if (entity.getFee() == null) {
            entity.setFee(0);
        }

        if (entity.getOffsetAmount() == null) {
            entity.setOffsetAmount(0);
        }

        if (entity.getTotalAmount() == null) {
            entity.setTotalAmount(0);
        }

        if (entity.getPaymentStatus() == null) {
            entity.setPaymentStatus(CustomerPaymentStatus.UNPAID);
        }

        entity.setNote(request.note());
    }

    public CustomerTransactionResponse toResponse(
            CustomerTransaction entity
    ) {
        return new CustomerTransactionResponse(
                entity.getId(),
                entity.getCustomerId(),
                entity.getTargetMonth(),

                DayRuleUtils.toResponse(
                        entity.getClosingDayType(),
                        entity.getClosingDayValue(),
                        entity.getClosingMonthOffset()
                ),

                DayRuleUtils.toResponse(
                        entity.getPaymentDayType(),
                        entity.getPaymentDayValue(),
                        entity.getPaymentMonthOffset()
                ),

                entity.getBillingAmount(),
                entity.getExpectedPaymentDate(),
                entity.getConfirmedPaymentDate(),
                entity.getPaidAmount(),
                entity.getFee(),
                entity.getOffsetAmount(),
                entity.getTotalAmount(),

                entity.getPaymentStatus() == null
                        ? CustomerPaymentStatus.UNPAID.name()
                        : entity.getPaymentStatus().name(),

                entity.getNote()
        );
    }

    private void applyClosingDayRule(CustomerTransaction entity, DayRule rule) {
        entity.setClosingDayType(rule == null ? null : rule.type());
        entity.setClosingDayValue(rule == null ? null : rule.value());
        entity.setClosingMonthOffset(rule == null ? 0 : rule.monthOffset());
    }

    private void applyPaymentDayRule(CustomerTransaction entity, DayRule rule) {
        entity.setPaymentDayType(rule == null ? null : rule.type());
        entity.setPaymentDayValue(rule == null ? null : rule.value());
        entity.setPaymentMonthOffset(rule == null ? 0 : rule.monthOffset());
    }
}

File: frontend/src/features/customer/mapper/customerTransactionMapper.ts
import type { DayRule } from '@/shared/types/dayRuleTypes'
import type {
  CustomerPaymentConfirmRequest,
  CustomerTransactionRequest,
  CustomerTransactionResponse,
} from '../types/customerApiTypes'
import type { CustomerPaymentConfirmPayload, CustomerTransaction } from '../types/customerTypes'

function toStringValue(value: string | null | undefined): string {
  return value ?? ''
}

function toNullableString(value: string | null | undefined): string | null {
  if (value == null || value === '') return null
  return value
}

function toDayRule(value: DayRule | null | undefined): DayRule | null {
  if (!value || !value.type) return null

  return {
    type: value.type,
    value: value.value ?? null,
    monthOffset: value.monthOffset ?? 0,
    label: value.label,
  }
}

function toDayRuleRequest(value: DayRule | null | undefined): DayRule | null {
  if (!value || !value.type) return null

  return {
    type: value.type,
    value: value.type === 'END_OF_MONTH' ? null : (value.value ?? null),
    monthOffset: value.monthOffset ?? 0,
  }
}

export function toCustomerTransaction(response: CustomerTransactionResponse): CustomerTransaction {
  return {
    id: response.id,
    customerId: response.customerId,
    deleteSelected: false,
    targetMonth: toStringValue(response.targetMonth),
    closingDayRule: toDayRule(response.closingDayRule),
    paymentDayRule: toDayRule(response.paymentDayRule),
    billingAmount: response.billingAmount,
    expectedPaymentDate: response.expectedPaymentDate,
    confirmedPaymentDate: response.confirmedPaymentDate,
    paidAmount: response.paidAmount,
    fee: response.fee,
    receivableAmount: (response.billingAmount ?? 0) + (response.fee ?? 0),
    offsetAmount: response.offsetAmount,
    totalAmount: response.totalAmount,
    paymentStatus: response.paymentStatus ?? 'UNPAID',
    note: toStringValue(response.note),
    _isNew: false,
    _isUpdated: false,
    _isDeleted: false,
  }
}

export function toCustomerTransactionRequest(row: CustomerTransaction): CustomerTransactionRequest {
  return {
    targetMonth: toNullableString(row.targetMonth),
    closingDayRule: toDayRuleRequest(row.closingDayRule),
    paymentDayRule: toDayRuleRequest(row.paymentDayRule),
    billingAmount: row.billingAmount,
    expectedPaymentDate: row.expectedPaymentDate,
    confirmedPaymentDate: row.confirmedPaymentDate,
    paidAmount: row.paidAmount,
    fee: row.fee,
    offsetAmount: row.offsetAmount,
    totalAmount: row.totalAmount,
    paymentStatus: row.paymentStatus,
    note: toNullableString(row.note),
  }
}

export function toCustomerPaymentConfirmRequest(
  payload: CustomerPaymentConfirmPayload,
): CustomerPaymentConfirmRequest {
  return {
    confirmedPaymentDate: payload.confirmedPaymentDate,
    paidAmount: payload.paidAmount,
    fee: payload.fee,
    offsetAmount: payload.offsetAmount,
    note: toNullableString(payload.note ?? null),
  }
}
